src/Streamly/Mem/Array.hs

```diff
@@ -271,9 +327,17 @@ splice arr1 arr2 = undefined
 -- first half retains values where the predicate is 'False' and the second half
 -- retains values where the predicate is 'True'.
 {-# INLINE partitionBy #-}
-partitionBy :: (a -> Bool) -> Array a -> (Array a, Array a)
-partitionBy f arr = undefined
+partitionBy :: (MonadIO m, Storable a) => (a -> Bool) -> Array a -> m (Array a, Array a)
```

Comment:
This should be an in-place mutation primitive. We should just swap the elements in the existing array and then slice the array into two arrays by sharing the same underlying array just different start/end indices.